Add Capabilities to RSDL (#441)
oasis-open/odata-rapid: docs/rsdl/rapid-rsdl-abnf.md

Reviewer: Where is the `paths` keyword/construct allowed within an RSDL file? Only at the end, or everywhere?

And how often can it appear? Once, or more than once? How are multiple occurences combined if they are allowed?

Reviewer: I propose to allow it only once, and only at the end of the file _after_ service and type declarations.

Author: agree.

structuredTypeMember = property / operation ; property, action, or function

property             = singlePropertyDefinition [ OWS (primitivePropertyCapabilities / singleNavigationCapabilities) ]
                     / collectionPropertyDefinition [ OWS ( collectionCapabilities / collectionNavigationCapabilities) ]

singlePropertyDefinition  = annotations [propertyModifier RWS] identifier OWS ":" OWS singleTypeReference

> collectionPropertyDefinition  = annotations [propertyModifier RWS] identifier OWS ":" OWS collectionTypeReference

Reviewer: ```suggestion
collectionPropertyDefinition  = annotations identifier OWS ":" OWS collectionTypeReference
```

> collectionPropertyDefinition  = annotations [propertyModifier RWS] identifier OWS ":" OWS collectionTypeReference

Reviewer: ```suggestion
collectionPropertyDefinition  = annotations identifier OWS ":" OWS collectionTypeReference
```

structuredTypeMember = property / operation ; property, action, or function

> property             = structuralProperty / referenceProperty

Reviewer: This is "semantic ABNF" because a parser can't know which kind it has encountered.

> property             = structuralProperty / referenceProperty

Author: I was playing with different ways to describe a navigation property. What do you think? If reference property makes more sense, we can change it throughout; if there's not a lot of value then we can stick with navigationProperty.

> property             = structuralProperty / referenceProperty

Reviewer: I‘d stick with navigationProperty.

structuredTypeMember = property / operation ; property, action, or function

property             = singlePropertyDefinition [ OWS (primitivePropertyCapabilities / singleNavigationCapabilities) ]
                     / collectionPropertyDefinition [ OWS ( collectionCapabilities / collectionNavigationCapabilities) ]

singlePropertyDefinition  = annotations [propertyModifier RWS] identifier OWS ":" OWS singleTypeReference

> collectionPropertyDefinition  = annotations [propertyModifier RWS] identifier OWS ":" OWS collectionTypeReference

Reviewer: If collection properties can't be key, we could remove the `[propertyModifier RWS]` part

Author: good catch!

            / nullableValuedSegment [ "/" castSegment ] [ RWS nullablePathCapabilities ]
            / collectionValuedSegment [ "/" castSegment ] [ RWS collectionPathCapabilities ]
            / singleRefSegment [ "/" castSegment ] [ RWS singleRefPathCapabilities ]
            / nullableRefSegment [ "/" castSegment ] [ RWS singleNullableReferenceCapability ]
            / collectionRefSegment [ "/" castSegment ] [ RWS ( collectionRefPathCapabilities / "/" keySegment RWS nullableRefPathCapabilities )]

serviceOperationPath = "/" singleValuedOperation [ "/" castSegment ] [ RWS singlePathCapabilities]
                     / "/" nullableValuedOperation [ "/" castSegment ] [ RWS nullablePathCapabilities]
                     / "/" collectionValuedOperation [ "/" castSegment ] [ RWS collectionPathCapabilities]
                     / "/" singleRefValuedOperation [ "/" castSegment ] [ RWS singleRefPathCapabilities]
                     / "/" nullableRefValuedOperation [ "/" castSegment ] [ RWS nullableRefPathCapabilities]
                     / "/" collectionRefValuedOperation [ "/" castSegment ] [ RWS collectionRefPathCapabilities]

> singleValuedSegment = identifier                   ; a single, non-nullable property
>                     / singleValuedOperation

Reviewer: `singleValuedOperation` is just a synonym for `identifier`, so this rule alternative doesn't add any value. Remove it and the synonym rule, and maybe just extend the comment:

```suggestion
singleValuedSegment = identifier                   ; a single, non-nullable property or single-valued operation
```
